### PR TITLE
Removed reference to binding in UserDefinedFunction

### PR DESCRIPTION
**Pull request by VamsiModem**

### CODE REVIEW

File: src/libraries/Microsoft.PowerFx.Core/Functions/UserDefinedFunctionLibrary.cs [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+
+namespace Microsoft.PowerFx.Core.Functions
+{
+    internal class UserDefinedFunctionLibrary
+    {
+        private TexlFunctionSet _library;
+
+        private readonly Dictionary<string, TexlBinding> _bindings;
+
+        public Dictionary<string, List<TexlFunction>>.KeyCollection FunctionNames => _library.FunctionNames;
```
Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.Core/Functions/UserDefinedFunctionLibrary.cs:22]: >string [](http://example.com/codeflow?start=26&length=6)

can you comment what 'string' is? 
function name? (with / without namespace?)

File: src/libraries/Microsoft.PowerFx.Core/Functions/UserDefinedFunction.cs [hunk 2]
```diff
@@ -30,21 +30,33 @@ internal class UserDefinedFunction : TexlFunction
     {
         private readonly bool _isImperative;
         private readonly IEnumerable<UDFArg> _args;
-        private TexlBinding _binding;
+        private readonly UserDefinedFunctionLibrary _library;
 
-        public override bool IsAsync => _binding?.IsAsync(UdfBody) ?? false;
+        public override bool IsAsync
```
Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.Core/Functions/UserDefinedFunction.cs:35]: >IsAsync [](http://example.com/codeflow?start=29&length=7)

is there any danger in this value being able to mutate on us?

File: src/libraries/Microsoft.PowerFx.Core/Functions/UserDefinedFunctionLibrary.cs (hunk 1 above)
Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.Core/Functions/UserDefinedFunctionLibrary.cs:22]: >Dictionary<string, List<TexlFunction>>.KeyCollection [](http://example.com/codeflow?start=15&length=52)

Can this be IEnumerable<string> ?

File: src/libraries/Microsoft.PowerFx.Core/Functions/UserDefinedFunctionLibrary.cs [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+            _bindings = new Dictionary<string, TexlBinding>();
+        }
+
+        public bool TryAdd(UserDefinedFunction function, TexlBinding binding)
+        {
+            Contracts.AssertValue(function);
+
+            if (_library.AnyWithName(function.Name))
+            {
+                return false;
```
Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.Core/Functions/UserDefinedFunctionLibrary.cs:38]: > return false; [](http://example.com/codeflow?start=15&length=14)

do we need to replace when the binding is updated?

File: src/libraries/Microsoft.PowerFx.Core/Functions/UserDefinedFunction.cs [hunk 4]
```diff
@@ -30,21 +30,33 @@ internal class UserDefinedFunction : TexlFunction
     {
         private readonly bool _isImperative;
         private readonly IEnumerable<UDFArg> _args;
-        private TexlBinding _binding;
+        private readonly UserDefinedFunctionLibrary _library;
```
Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.Core/Functions/UserDefinedFunction.cs:33]: >UserDefinedFunctionLibrary [](http://example.com/codeflow?start=25&length=26)

this adds a level of indirection, but can we remove the dependency entirely?
Ie, if we're just doing intellisense at the callsite, we don't need this. 
And if we're doing codegen , we should already have it through a separate channel. 

(this begs the issue for IsAsync)

Comment (VamsiModem) [reply src/libraries/Microsoft.PowerFx.Core/Functions/UserDefinedFunction.cs:33]: Makes sense, the only reason we have reference binding/library is to derive the value for IsAsync as it is used in several places in the PAClient.

**Comment by LucGenetier** [comment]
✅ No public API change.

**Comment by LucGenetier** [comment]
✅ No public API change.

File: src/libraries/Microsoft.PowerFx.Core/Functions/UserDefinedFunction.cs (hunk 2 above)
Comment (VamsiModem) [reply src/libraries/Microsoft.PowerFx.Core/Functions/UserDefinedFunction.cs]: we are not deriving IsAsync from Binding or library anymore.

**Comment by LucGenetier** [comment]
✅ No public API change.

**Pull request closed without merging**